skimage/feature/_canny.py

```diff
@@ -265,6 +269,16 @@ def fsmooth(x):
 
         high_threshold = np.percentile(magnitude, 100.0 * high_threshold)
         low_threshold = np.percentile(magnitude, 100.0 * low_threshold)
+    else:
+        #
+        # Else if high_threshold and/or low_threshold are callables
+        # call them to determine the threshold value / image
+        #
+        if callable(high_threshold):
+            high_threshold = high_threshold(magnitude)
+        if callable(low_threshold):
+            low_threshold = low_threshold(magnitude)
```

Comment:
I'm not going to hold up the PR for this, but ideally we would do some sanity checking here: use try/except so that if there's an error, *we* raise a TypeError with the appropriate error message (the input type of the callable passed to Canny is wrong; and reproduce the traceback that came with the call), and if the type of high_threshold and low_threshold are wrong, then we raise a TypeError saying that the *output* type of the callable is wrong.

If you can add this we would be eternally grateful! =)

Reply:
Just working on the example quickly over lunch ;)